SDK quickstart structure (done on NodeJS) (#373)
permitio/docs: docs/sdk/quickstart-by-language/quickstart-nodejs.mdx

can do what, based on the policy that you defined.

Each user, once his identity is verified by the authentication provider, gets issued a session. During this session, his unique id
obtained from the JWT (JSON Web Token), which is synced to Permit, maps to a specific role within the Policy.

Each role has a set of rules that were pre-defined, which a policy engine can understand, evaluate, and therefore return if the user
is permitted or not, based on the user id, action and resource passed into the SDK function.

You can run a permission check with permit.check(), passing in 3 arguments:

- `user`: a unique string id that identifies the user doing the action (and the currently logged in user).
- `action`: the action performed.
- `resource`: the resource (object) the action is performed on.

In the following example we are checking that a user with the unique id `john@smith.com` can `read` a `document`.

```javascript
const permitted = await permit.check("john@permit.io", "read", "document");

if (permitted) {
  console.log("John is PERMITTED to read a document");
} else {
  console.log("John is NOT PERMITTED to read a document");
}
```

> ## Other basic functionality with the SDK

Reviewer: As this page shouldn't be a remix of the **Permit Product Quickstart** tutorial, I think this `fundamental` and `basic` should be "flatly" categorized into the various main functions of the **Permit client**:
- Check (enforcement)
- API (configuration)
Beside of that, we should also dedicate a section for unique SDK functions, such as `elements` that we have in the JS SDK.

> In this quickstart, we'll introduce you to some core Permit.io functionality using the SDK. As you'll see, the SDK is designed to

Reviewer: As the previous comment, we should have more general intro that say something like `the API support let's you call various API functions, for example, the following are some of the most-usage functions`.

Reviewer: Not sure how did you address this comment

Reviewer: Seems like this comment is missed in the PR again

Author: I've addressed this by defining sections, I'm not sure what else you mean? The third comment you left on this, is on outdated code that does not exist anymore, so I'm confused if you are referring to the previous iteration, or you missed the changes while reviewing again. I've sent you a message on Slack to explain further.

be simple and logical, making it easy to develop and manage policies.

### Creating a resource

A resource is the fundamental building block of a policy. It's what the user can interact with. It could be a 'document', a 'building'
or even a 'patient health record'.

```javascript

> await permit.api.resource.create({

Reviewer: In this case, I would add the resource instance action before the resource itself. The reason is the cadence of using such API calls via the API. We should focus more on the `data` operations and less on the `control` operations. We could have one for each plan, starting from the data and moving up.
- Sync user (very important and relevant)
- Role assignment
- Resource creation
- Copy environment

Author: Actually, this answered by above question

  key: "repository",
  name: "Repository",
  actions: {
    clone: {},
    push: {},
    view: {},
  },
});
```

### Creating a new role with permissions

A role defines a set of resources and the specific actions a user can perform on those resources. A user can only perform these
actions if they have been assigned the corresponding role.

````javascript
await permit.api.roles.create({
    key: "manager",
    name: "Manager",
    permissions: [
      "repository:clone",
      "repository:view",
    ],
  });```
````

> ### Assigning a role

Reviewer: Role assignment is a more common use case than creating a resource/role.

Author: Do you just want me to delete it? I've also talked about role assignment below, but the idea here was that usually with a basic RBAC policy, you create the resource, then you create a role (also checking the permissions) and finally assign the role to the user. It's just the logical flow IMO

A role assignment, as the name suggests, assigns a specific role with defined permissions to a user identity. This role assignment
is specific to a particular tenant.

```javascript
await permit.api.roleAssignments.assign({
  user,
  role,
  tenant: "default",
});
```

## Working with the basic SDK functionality

> For a comprehensive overview of basic SDK functionality, please refer to the essential list of SDK functions available

Reviewer: 1. That could be an admonition in the previous proposed section that refers to the function support.
2. The link here should be to the feature compatibility page. We decided to remove the mess of different pages for each SDK.
3. We need to have some form of conclusion that refers to the SDK reference docs. In JS case, to this one - https://permitio.github.io/permit-node/classes/Permit.html

[here](/category/nodejs).

## Working with the full SDK specification

> As part of your IDE (Integrated Development Environment), we highly suggest that you install an extension (such as **Tabnine** or

Reviewer: Code completion (or, as VSCode calls it, IntelliSense) is a feature that examines the AST of the project dependencies and proposes various options for each SDK the project uses. GitHub copilot is a product that uses AI for better completion where the traditional lacks. https://en.wikipedia.org/wiki/Code_completion

We shouldn't propose using Copilot as a recommendation for working with the SDK for two reasons:
1. It means our SDK lacks structure (which is not the case), and that it is not intuitive.
2. Copilot can propose SDK paths that do not exist or work at all.
See also the comment below regarding the whole section.

  It auto-completes code snippets and suggests the next steps, speeding up the development process.

- `Error Reduction`: By suggesting correct function names and parameters, IntelliSense helps avoid common typos and syntax errors,
  ensuring that your code runs smoothly.

- `Learning Aid`: IntelliSense acts as an interactive guide, showing you available methods and properties as you type. This is especially
  beneficial when working with a new SDK, as it helps you explore and understand the SDK's capabilities without constantly referring to
  external documentation.

- `Contextual Awareness`: IntelliSense provides suggestions based on the current context of your code, making it easier to find relevant
  functions and methods quickly. This contextual awareness streamlines coding, particularly in complex projects with multiple SDKs.
- `Comprehensive Documentation`: Integrated documentation snippets and parameter information make it easier to understand what each
  function does and how to use it effectively. This immediate access to information enhances comprehension and implementation

>   efficiency.

Reviewer: All this part could be a nice article about code completion features, but it has no reason to be included in every SDK page.

```javascript
import { Permit } from "permitio";
```

### 3. Create a new Instance of the SDK

Next, create a new Permit object instance of the Permit SDK. You'll need a secret API key, which you can obtain by following the instructions
provided in the setup section of the Permit.io documentation.

```javascript
// This first line initializes the SDK and connects your Node.js app
// to the Permit.io PDP container you've set up in the previous step.
const permit = new Permit({
  // your API Key
  token: "[YOUR_API_KEY]",

  // in production, you might need to change this url to fit your deployment
  pdp: "http://localhost:7766",

  // if you want the SDK to emit logs, uncomment this:
  // log: {
  //   level: "debug",
  // },

  // The SDK returns false if you get a timeout / network error
  // if you want it to throw an error instead, and let you handle this, uncomment this:
  // throwOnError: true,
});
```

Replace **`YOUR_API_KEY`** with the actual API key you obtained. Find instructions how to fetch your API
key [here](/overview/connecting-your-app#1-get-your-permit-environment-api-key).

## The SDK enforcement functionality - Checking for permissions

> `permit.check()` is the fundamental SDK function that controls the enforcement in your application level code. It acts as a guard of who

Reviewer: ```suggestion
`permit.check()` is the fundamental SDK function that controls the permissions enforcement in your application code. It acts as a guard of who
```

can do what, based on the policy that you defined.

> If you would like to learn more about the `permit.check()` function, and how it works with different policy models, you can read
> more about it [here](/how-to/enforce-permissions/check).

Reviewer: ```suggestion
:::tip
If you would like to learn more about the `permit.check()` function, and how it works with different policy models, you can read
more about it [here](/how-to/enforce-permissions/check).
:::
```

You can run a permission check with permit.check(), passing in 3 arguments:

> - `user`: a unique string id that identifies the user performing the action.

Reviewer: If we are showing the arguments in this document, we should also show the various types of the user/resource argument types. User could be an object, not only a string

await permit.api.resource.create({
  key: "repository",
  name: "Repository",
  actions: {
    clone: {},
    push: {},
    view: {},
  },
});
```

### Creating a new role with permissions

A role defines a set of resources and the specific actions a user can perform on those resources. A user can only perform these
actions if they have been assigned the corresponding role.

````javascript
await permit.api.roles.create({
    key: "manager",
    name: "Manager",
    permissions: [
      "repository:clone",
      "repository:view",
    ],
  });```
````

### Syncing a user to Permit

The sync user function allows you to take an existing user from your authentication solution, and add them inside of Permit.

```javascript
await permit.api.syncUser({
  key: "key",
  email: "example@permit.io",
  first_name: "John",
  last_name: "Smith",
  attributes: {},
});
```

### Assigning a role

A role assignment, as the name suggests, assigns a specific role with defined permissions to a user identity. This role assignment
is specific to a particular tenant.

```javascript
await permit.api.roleAssignments.assign({
  user,
  role,
  tenant: "default",
});
```

## Unique SDK functions

:::note

> For a comprehensive overview of basic SDK functionality, please refer to the [compatibility list](/sdk/sdks-overview) of SDK

Reviewer: ```suggestion
For a comprehensive overview of the various SDK functions, please refer to the [compatibility list](/sdk/sdks-overview) of SDK
```

  // log: {
  //   level: "debug",
  // },

  // The SDK returns false if you get a timeout / network error
  // if you want it to throw an error instead, and let you handle this, uncomment this:
  // throwOnError: true,
});
```

Replace **`YOUR_API_KEY`** with the actual API key you obtained. Find instructions how to fetch your API
key [here](/overview/connecting-your-app#1-get-your-permit-environment-api-key).

## The SDK enforcement functionality - Checking for permissions

`permit.check()` is the fundamental SDK function that controls the permissions enforcement in your application code. It acts as a guard of who
can do what, based on the policy that you defined.

:::tip
If you would like to learn more about the `permit.check()` function, and how it works with different policy models, you can read
more about it [here](/how-to/enforce-permissions/check).
:::

You can run a permission check with permit.check(), passing in 3 arguments:

- `user`: a unique string id or object that identifies the user performing the action.
- `action`: the action performed.
- `resource`: the resource (object) the action is performed on. Can be a string or an object.

> :::tip

Reviewer: If we mentioning ABAC in this way, then we should also mention the ReBAC convention using the tuple convention of `:`

Reviewer: I personally think it could be a much smaller comment about the argument types and do not necessarily mention ABAC/ReBAC

Author: This is an opinionated comment and I personally disagree, when giving examples of other models implementations, if we mention one model, I don't think there is necessary a need to always mention all the other models. ReBAC is a far more complex policy model to understand, and it's not relevant to be mentioning it's implementation in the quickstart. 

Initially you wanted me to mention ABAC and how the check function can change and be different, now you don't want me to mention ABAC. Can you give me an example how you want me to talk about mentioning other argument types (which is ABAC) without a reference to ABAC?

Reviewer: I didn't express myself well in the first comment. I'll try again then.

The check function is not supposed to explain RBAC, ABAC, or ReBAC. The only thing we could explain is the different possible arguments. In our check function, there are three types of arguments—string for type, string for instance reference, and object (that could include some data).

I read my comment again and understood that I forgot to put an `etc.` after the example of the user as an object.  I didn't mean to mention ABAC in the first comment.

What I'm asking is that you mention briefly that the check function could have other arguments and refer to the main `check` page documentation.

```javascript
await permit.api.resource.create({
  key: "repository",
  name: "Repository",
  actions: {
    clone: {},
    push: {},
    view: {},
  },
});
```

### Creating a new role with permissions

A role defines a set of resources and the specific actions a user can perform on those resources. A user can only perform these
actions if they have been assigned the corresponding role.

````javascript
await permit.api.roles.create({
    key: "manager",
    name: "Manager",
    permissions: [
      "repository:clone",
      "repository:view",
    ],
  });```
````

### Syncing a user to Permit

The sync user function allows you to take an existing user from your authentication solution, and add them inside of Permit.

```javascript
await permit.api.syncUser({
  key: "key",
  email: "example@permit.io",
  first_name: "John",
  last_name: "Smith",
  attributes: {},
});
```

### Assigning a role

A role assignment, as the name suggests, assigns a specific role with defined permissions to a user identity. This role assignment
is specific to a particular tenant.

```javascript
await permit.api.roleAssignments.assign({
  user,
  role,
  tenant: "default",
});
```

> ## Unique SDK functions

Reviewer: This title is redundant (especially because the section contains a single admonition

:::note
For a comprehensive overview of the various SDK functions, please refer to the [compatibility list](/sdk/sdks-overview) of SDK
functions available per language.
:::

> 

Reviewer: I'm still missing here the component of the other pages under the Node.js part

Author: By component do you mean tiles? What component? Please be more specific of your suggestions/requirements
